### DISCUSSION TITLE
In a workspace, how to list packages that exist internally in the repository?

### DISCUSSION THREAD

**Discussion by kenrick95** [pnpm/pnpm#4150] [Q&A] [answered]
So in a lerna monorepo, I could do something like 

```
lerna list --all
```

and it will return a list of all internal packages in the monorepo (including private ones), as specified from `lerna.json`. This list returns the following information:

- package name
- package version
- package location (relative to the monorepo root)
- is it private or not?

-----

I wonder what's the equivalent one for pnpm workspace?

I've tried:

```
pnpm list
```

This list down all dependencies of the root package.json and this includes the external ones.

```
pnpm list --recursive --depth 1
```

```
pnpm list --recursive --depth 0
```

This seem to be closer to what I want as it returns list of packages plus their dependencies including external ones. But this command is really slow in a big workspace and I don't really need the dependency information

**Comment by zkochan** [comment, marked as answer]
```
pnpm list --recursive --depth -1
```

**Reply by kenrick95** [reply]
oh wow, sorry I misread the documentation! Just realized it is a minus one instead of one 😅 Thanks a lot for the quick reply

**Reply by binjospookie** [reply]
Hi!

Is it possible to output list in topological order?

Like `lerna list --toposort `

**Reply by zkochan** [reply]
Currently there is no such feature

**Reply by binjospookie** [reply]
As far as I understand, pnpm runs scripts in workspace in topological order. That is, the functionality is implemented, but there is no access to it from the outside.

Am I right?

**Reply by zkochan** [reply]
You are right

**Reply by binjospookie** [reply]
Is it possible to add such an option to the list command? 

I would be happy to help, but this is my first dive into pnpm source code :c

**Reply by zkochan** [reply]
Should be possible.

**Reply by aleclarson** [reply]
`pnpm list --toposort` would be great

**Reply by aleclarson** [reply]
i thought maybe `pnpm -r exec sh -c 'echo $PNPM_PACKAGE_NAME'` would print in topological order, but it doesn't seem to.
